prometheus/go_collector_go117.go

```diff
@@ -0,0 +1,321 @@
+
+	// rm... fields all pertain to the runtime/metrics package.
+	rmSampleBuf []metrics.Sample
+	rmSampleMap map[string]*metrics.Sample
+	rmDescs     []*Desc
+	rmValTypes  []ValueType
+
+	// With Go 1.17, the runtime/metrics package was introduced.
+	// From that point on, metric names produced by the runtime/metrics
+	// package could be generated from runtime/metrics names. However,
+	// these differ from the old names for the same values.
+	//
+	// This field exist to export the same values under the old names
+	// as well.
+	msMetrics memStatsMetrics
+}
+
+// NewGoCollector is the obsolete version of collectors.NewGoCollector.
+// See there for documentation.
+//
+// Deprecated: Use collectors.NewGoCollector instead.
+func NewGoCollector() Collector {
+	descriptions := metrics.All()
+
+	// Generate a Desc and ValueType for each runtime/metrics metric.
+	descs := make([]*Desc, 0, len(descriptions))
+	valTypes := make([]ValueType, 0, len(descriptions))
+	sampleBuf := make([]metrics.Sample, 0, len(descriptions))
+	sampleMap := make(map[string]*metrics.Sample, len(descriptions))
+	for i := range descriptions {
+		d := &descriptions[i]
+		name, ok := nameFromRuntimeMetrics(d)
+		if !ok {
+			// TODO: Log this somewhere?
+			continue
+		}
+
+		// Set up sample buffer for reading, and a map
+		// for quick lookup of sample values.
+		sampleBuf = append(sampleBuf, metrics.Sample{
+			Name: d.Name,
+		})
+		sampleMap[d.Name] = &sampleBuf[len(sampleBuf)-1]
+
+		descs = append(descs, NewDesc(
+			name,
+			d.Description,
+			nil, nil,
+		))
+
+		valType := GaugeValue
+		if d.Cumulative {
+			valType = CounterValue
+		}
+		valTypes = append(valTypes, valType)
+	}
+	return &goCollector{
+		base:        newBaseGoCollector(),
+		rmSampleBuf: sampleBuf,
+		rmSampleMap: sampleMap,
+		rmDescs:     descs,
+		rmValTypes:  valTypes,
+		msMetrics:   goRuntimeMemStats(),
+	}
+}
+
+// Describe returns all descriptions of the collector.
+func (c *goCollector) Describe(ch chan<- *Desc) {
+	c.base.Describe(ch)
+	for _, i := range c.msMetrics {
+		ch <- i.desc
+	}
+	for _, desc := range c.rmDescs {
+		ch <- desc
+	}
+}
+
+// Collect returns the current state of all metrics of the collector.
+func (c *goCollector) Collect(ch chan<- Metric) {
+	// Collect base non-memory metrics.
+	c.base.Collect(ch)
+
+	// Populate runtime/metrics sample buffer.
+	metrics.Read(c.rmSampleBuf)
+
+	for i, sample := range c.rmSampleBuf {
+		desc := c.rmDescs[i]
+		switch sample.Value.Kind() {
+		case metrics.KindUint64:
+			ch <- MustNewConstMetric(desc, c.rmValTypes[i], float64(sample.Value.Uint64()))
+		case metrics.KindFloat64:
+			ch <- MustNewConstMetric(desc, c.rmValTypes[i], sample.Value.Float64())
+		case metrics.KindFloat64Histogram:
+			h := newConstRuntimeMetricsHistogram(desc, sample.Value.Float64Histogram())
+			if sum, ok := c.exactSumFor(sample.Name); ok {
+				h.exactSum(sum)
+			}
+			ch <- h
+		case metrics.KindBad:
+			// TODO: Unsupported metric. Log this somewhere?
```

Comment:
Hm.. this is scary. We might want to introduce some errReporter logic so user can choose how to be informed (by default ignore)

Reply:
... and it tests we should check that for sure!(not ignore, but panic or return err)

Reply:
I added a panic here, skipped collection of these metrics (so the panic shouldn't fire), and added a test that allows for auditing of these cases for a new Go version.